Replace explicit library home/end key handlers with a moveCursor override

robbert-vdh opened the pull request:
By simply overriding the cursor movement function instead of adding an explicit key event the home and end keys in the library table view now interact as expected with other hotkeys and cursor actions. <kbd>Shift+End</kbd> now select all items through the end of the list, and <kbd>Ctrl+Home</kbd> jumps to the first item without selecting it so it can be toggled on and off with Space.

Tagging @ronso0 since they suggested it in #11090. Sorry that it took so long (since making this actual change only took a minute heh). Had a busy weekend. Once either of these PRs are merged I'll rebase the other one and refactor the `moveCursor()` function a bit.

robbert-vdh committed:
Replace library home/end key handlers with cursors

By simply overriding the cursor movement function instead of adding an
explicit key event the home and end keys in the library table view now
interact as expected with other hotkeys and cursor actions. Shift+End
now select all items through the end of the list, and Ctrl+Home jumps to
the first item without selecting it so it can be toggled on and off with
Space.

ronso0 commented:
Thank you, LGTM

It's just I'd prefer to have all keyhandling methods in one class, probably WTrackTableView.
@robbert-vdh Do you mind moving `moveCursor()` there, somewhere near `keyPressEvent`?

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -260,3 +260,35 @@ void WLibraryTableView::focusInEvent(QFocusEvent* event) {
+
+QModelIndex WLibraryTableView::moveCursor(CursorAction cursorAction,
+        Qt::KeyboardModifiers modifiers) {
+    // By default the home and end keys move to the first and last column rather
+    // than the first and last row
```
I had to read that a few times to figure you describe what is _overridden_, not the new implementation. Let's rephrase it:
```suggestion
    // Make the home and end keys move to the first and last row rather than
    // the first and last column (QAbstractItemView default)
```

robbert-vdh replied:
Sure, I'll change that. Btw, I'm not the only one who does this ha. if you grep for `By default` you'll get 46 other matches.

ronso0 commented:
> ...  if you prefer that instead of merging the classes. 

I'd prefer merging, but as I wrote in #11090 I'd love to hear an opinion from an 'older' team member on that topic before anyone spends time on this.

robbert-vdh committed:
Change wording in home/end moveCursor override

As per
https://github.com/mixxxdj/mixxx/pull/11100#discussion_r1040296239.

robbert-vdh commented:
Oh and the focus in event handler is also overridden in `WLibraryTableView`. Should I move just `moveCursor()`? Or move both (sounds like that would be out of scope for this PR)? Or will you move them after this is merged?

robbert-vdh commented:
Rebased on top of the changes from the just merged #11090.

ronso0 reviewed (commented):
LGTM, works as expected. Just some comments.

ronso0 commented:
I merged this into my personal branch and experienced some issues last night with mixed keyboard and controller navigation but couldn't investigate further at that time:
* scroll around with `[Library], MoveVertical`
* select a track
* hold Shift
* press Home or End

would not give my the expected selection but less.
I'll try to reproduce again soonish.

robbert-vdh committed:
Reorder up/down library cursor handler fallback

As suggested in
https://github.com/mixxxdj/mixxx/pull/11100#discussion_r1044904901.

robbert-vdh commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -263,50 +263,80 @@ void WLibraryTableView::focusInEvent(QFocusEvent* event) {
+                    return pModel->index(0, column);
                 } else {
                     return pModel->index(pModel->rowCount() - 1, column);
                 }
```
Done.

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -263,50 +263,80 @@ void WLibraryTableView::focusInEvent(QFocusEvent* event) {
+        // or the user needs to reach for the mouse or keyboard when moving
+        // between 12/C#m/E and 1/G#m/B. This is very similar to
+        // `moveSelection()`, except that it doesn't actually modify the
+        // selection
```
:+1: yeah, that's what I mean. Worth noting I'd say

robbert-vdh commented:
> * scroll around with `[Library], MoveVertical`
> * select a track
> * hold Shift
> * press Home or End

I just tried this and naively moving the cursor with a my controller and then pressing Shift+Home on my keyboard seems to work as expected. I'll see if I can find a situation where it doesn't.

robbert-vdh committed:
Clarify moveCursor not having side effects

daschuer reviewed (approved):
This works and looks good.
Thank you.

I have just noticed the inversion of the selection when crossing the list boundaries. 
Not sure if this is worth a fix. At least this is unrelated to this PR. 

@ronso0 do you want to do more tests or reviews before merge?

daschuer reviewed (approved):
Still looks good, thank you.

daschuer commented:
@ronso0  Did you find time for the double check?

ronso0 commented:
I didn't explicitly check (tbh I forgot..) but seems I had this already merged into my personal branch and I didn't encounter issues last times I used that live.

Thanks @robbert-vdh for this!
